ynr/apps/elections/views.py

```diff
@@ -139,821 +125,51 @@ def dispatch(self, *args, **kwargs):
 
     def get_context_data(self, **kwargs):
-        from candidates.election_specific import shorten_post_label
 
         context = super().get_context_data(**kwargs)
 
-        context["ballot"] = get_object_or_404(
+        context["ballot"] = ballot = get_object_or_404(
             Ballot.objects.all().select_related("post", "election"),
             ballot_paper_id=context["election"],
         )
 
-        mp_post = context["ballot"].post
-        context["election"] = election = context["ballot"].election
-        context["post_id"] = post_id = mp_post.slug
-        context["post_obj"] = mp_post
-
-        documents_by_type = {}
-        # Make sure that every available document type has a key in
-        # the dictionary, even if there are no such documents.
-        doc_lookup = {
-            t[0]: (t[1], t[2]) for t in OfficialDocument.DOCUMENT_TYPES
-        }
-        for t in doc_lookup.values():
-            documents_by_type[t] = []
-        documents_for_post = OfficialDocument.objects.filter(
-            ballot=context["ballot"]
-        )
-        for od in documents_for_post:
-            documents_by_type[doc_lookup[od.document_type]].append(od)
-        context["official_documents"] = documents_by_type.items()
-        context["some_official_documents"] = documents_for_post.count()
-
-        context["post_label"] = mp_post.label
-        context["post_label_shorter"] = shorten_post_label(
-            context["post_label"]
+        context["candidates"] = Membership.objects.memberships_for_ballot(
+            ballot
         )
 
-        context["redirect_after_login"] = context["ballot"].get_absolute_url()
+        try:
+            context["sopn"] = ballot.sopn
+        except OfficialDocument.DoesNotExist:
+            context["sopn"] = None
 
-        context["post_data"] = {"id": mp_post.slug, "label": mp_post.label}
-
-        ballot = context["ballot"]
-
-        context["candidates_locked"] = ballot.candidates_locked
-
-        context["has_lock_suggestion"] = SuggestedPostLock.objects.filter(
-            ballot=ballot
-        ).exists()
-
-        if self.request.user.is_authenticated:
-            context[
-                "current_user_suggested_lock"
-            ] = SuggestedPostLock.objects.filter(
-                user=self.request.user, ballot=ballot
-            ).exists()
-
-        context["suggest_lock_form"] = SuggestedPostLockForm(
-            initial={"ballot": ballot}
+        context["membership_edits_allowed"] = ballot.user_can_edit_membership(
+            self.request.user
         )
 
-        if self.request.user.is_authenticated:
+        if ballot.has_lock_suggestion:
             context[
-                "user_has_suggested_lock"
-            ] = SuggestedPostLock.objects.filter(
-                user=self.request.user, ballot=ballot
+                "current_user_suggested_lock"
+            ] = ballot.suggestedpostlock_set.filter(
+                user=self.request.user
```

Comment:
This is throwing `'AnonymousUser' object is not iterable` if I'm not logged in. Can you reproduce?

Reply:
Hmm, no I can't. Can you give me more information about the set up?

Reply:
Are you sure you're testing it with a ballot where `ballot.has_lock_suggestion == True`?

In fact, was it intentional to completely replace `self.request.user.is_authenticated` with `ballot.has_lock_suggestion` a few lines up?

here's a stack trace if that helps:

```
Traceback (most recent call last):
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/chris/repositories/github_chris48s/yournextrepresentative/ynr/apps/elections/helpers.py", line 35, in __call__
    return view(request, *args, **kwargs)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/views/generic/base.py", line 71, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/utils/decorators.py", line 45, in _wrapper
    return bound_method(*args, **kwargs)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/views/decorators/cache.py", line 31, in _cache_controlled
    response = viewfunc(request, *args, **kw)
  File "/home/chris/repositories/github_chris48s/yournextrepresentative/ynr/apps/elections/views.py", line 125, in dispatch
    return super().dispatch(*args, **kwargs)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/views/generic/base.py", line 97, in dispatch
    return handler(request, *args, **kwargs)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/views/generic/base.py", line 158, in get
    context = self.get_context_data(**kwargs)
  File "/home/chris/repositories/github_chris48s/yournextrepresentative/ynr/apps/elections/views.py", line 153, in get_context_data
    user=self.request.user
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/db/models/manager.py", line 82, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/db/models/query.py", line 892, in filter
    return self._filter_or_exclude(False, *args, **kwargs)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/db/models/query.py", line 910, in _filter_or_exclude
    clone.query.add_q(Q(*args, **kwargs))
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/db/models/sql/query.py", line 1290, in add_q
    clause, _ = self._add_q(q_object, self.used_aliases)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/db/models/sql/query.py", line 1318, in _add_q
    split_subq=split_subq, simple_col=simple_col,
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/db/models/sql/query.py", line 1224, in build_filter
    self.check_related_objects(join_info.final_field, value, join_info.opts)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/db/models/sql/query.py", line 1086, in check_related_objects
    for v in value:
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/utils/functional.py", line 257, in inner
    return func(self._wrapped, *args)
TypeError: 'AnonymousUser' object is not iterable
```

Reply:
I still can't reproduce this, but I've pulled the code out in to its own method now for easy of reading. This means the query will never get call for non-authenticated users anyway.